### ISSUE TITLE
Non-negative conjugate gradient?

### ISSUE THREAD

**Issue by erikfrey** [google/jaxopt#294]
We are using jaxopt to solve Ax = b s.t. x >= 0 using ProjectedGradient:

```
pg = jaxopt.ProjectedGradient(objective, projection_non_negative, ...
```

However our matrix A is positive definite, so I would love to use conjugate gradient.  In some testing, using jax.scipy.sparse.linalg.cg doubles our pipeline performance, which is great, but then we lose the non-negative constraint.  Googling around I found this:

https://github.com/david-cortes/nonneg_cg

I was going to try it out in JAX, but before I do: is this something that would be useful to jaxopt?  Or is there a way to do non-negative conjugate gradient in jaxopt already that I haven't noticed?  Thanks!

**Comment by mblondel** [comment]
It seems like the paper "A conjugate gradient type method for the nonnegative constraints optimization problems" is a bit low on the citation number side (only 5 citations). IMO, it's better to focus on more established, standard algorithms. @fabianp may also have an opinion on this particular case.

**Comment by erikfrey** [comment]
For sure, defer to you guys as to whether this makes sense for jaxopt.

In case it's useful to know, non-negative conjugate constraints are useful in physics simulation, where you have very large constraint jacobian matrices that determine collision forces and other things.  MuJoCo has implemented non-negative cg:

https://github.com/deepmind/mujoco/blob/main/src/engine/engine_solver.c#L710

And we plan to implement this in JAX for our simulator, [Brax](https://github.com/google/brax).

So the approach may not have many research citations but it is useful, and in practice converges much faster than other approaches, for our domain at least.  But totally get if this isn't appropriate for jaxopt, feel free to close the issue if so.  Thanks!

**Comment by mblondel** [comment]
If the code is not too large and it's confirmed that it works better for some use cases, then I would be open to it. We may want to add (vanilla) CG directly to JAXopt too. Indeed our API gives access to individual iterations, which is sometimes useful.

**Comment by erikfrey** [comment]
OK after some tinkering I was able to get `ProjectedGradient` to be just as fast as `jax.scipy.sparse.linalg.cg`.  Two things I had to do:

* set `stepsize` > 0 to disable the backtracking line search
* set `implicit_diff`= False, this also had a pretty significant impact on perf

I don't yet have a firm grasp as to why these two changes had such a dramatic impact, or why my use case is different from the average use case for which jaxopt has determined those default parameters - maybe in some future iteration of jaxopt documentation, you could include some guidance about when it is appropriate to change these default parameters?

But for now at least I'm happy with the performance of `ProjectedGradient` and probably won't pursue adapting non-negative cg.

Thanks for the great work in jaxopt!

**Comment by mblondel** [comment]
Thanks a lot @erikfrey for the heads up. Happy to hear that you managed to reach a satisfactory level of performance. I agree we could do a better job at explaining what are the defaults and how to change them.
